politeiawww/mail/client.go

```diff
@@ -0,0 +1,259 @@
+//
+// This function satisfies the Mailer interface.
+func (c *client) IsEnabled() bool {
+	return !c.disabled
+}
+
+// SendTo sends an email to a list of recipient email addresses.
+// This function does not rate limit emails and a recipient does
+// does not need to correspond to a politeiawww user. This function
+// can be used to send emails to sysadmins or similar cases.
+//
+// This function satisfies the Mailer interface.
+func (c *client) SendTo(subject, body string, recipients []string) error {
+	if c.disabled || len(recipients) == 0 {
+		return nil
+	}
+
+	// Setup email
+	msg := goemail.NewMessage(c.mailAddress, subject, body)
+	msg.SetName(c.mailName)
+
+	// Add all recipients to BCC
+	for _, v := range recipients {
+		msg.AddBCC(v)
+	}
+
+	return c.smtp.Send(msg)
+}
+
+// SendToUsers sends an email to a list of recipient email
+// addresses. The recipient MUST correspond to a politeiawww user
+// in the database for the email to be sent. This function rate
+// limits the number of emails that can be sent to any individual
+// user over a 24 hour period. If a recipient is provided that does
+// not correspond to a politeiawww user, the email is simply
+// skipped. An error is not returned.
+//
+// This function satisfies the Mailer interface.
+func (c *client) SendToUsers(subjects, body string, recipients map[uuid.UUID]string) error {
+	if c.disabled || len(recipients) == 0 {
+		return nil
+	}
+
+	filtered, err := c.filterRecipients(recipients)
+	if err != nil {
+		return err
+	}
+
+	// Handle valid recipients.
+	err = c.SendTo(subjects, body, filtered.valid)
+	if err != nil {
+		return err
+	}
+
+	// Handle invalid recipients.
+	err = c.SendTo(limitEmailSubject, limitEmailBody, filtered.invalid)
+	if err != nil {
+		return err
+	}
+
+	// Update email histories on db.
+	err = c.mailerDB.EmailHistoriesSave(filtered.histories)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// filteredRecipients contains the filtered recipients divided into valid,
+// those that are able to receive emails, and invalid, those that have hit
+// the email rate limit, but have not yet received the warning email. It will
+// also contains the updated email histories for each user email present on the
+// valid and invalid lists.
+type filteredRecipients struct {
+	valid     []string
+	invalid   []string
+	histories map[uuid.UUID]user.EmailHistory
+}
+
+// filterRecipients filters the users map[userid]email argument into the
+// filteredRecipients struct.
+func (c *client) filterRecipients(users map[uuid.UUID]string) (*filteredRecipients, error) {
+	// Compile user IDs from recipients and get their email histories.
+	ids := make([]uuid.UUID, 0, len(users))
+	for id := range users {
+		ids = append(ids, id)
+	}
+	hs, err := c.mailerDB.EmailHistoriesGet(ids)
+	if err != nil {
+		return nil, err
+	}
+
+	// Divide recipients into valid and invalid recipients, and parse their
+	// new email history.
+	var recipients filteredRecipients
```

Comment:
Initializing a struct like this when the struct contains fields that have nil zero values (slices/maps) greatly increases the chance that you introduce a bug where the struct is returned with one of the fields as nil instead of an empty slice/map and causes a panic in the calling function when it tries to access the field because it expects it to be filled in.

A better way to do it is the following.

```
	// Divide recipients into valid and invalid recipients, and parse their
	// new email history.
	var (
		valid     = make([]string, 0, len(users))
		invalid   = make([]string, 0, len(users))
		histories = make(map[uuid.UUID]user.EmailHistory, len(users))
	)

	// Code that populates valid, invalid, and histories
	// ....
	// ...

	return &filteredRecipients{
		Valid:     valid,
		Invalid:   invalid,
		Histories: histories,
	}, nil
```

This is not only less prone to bugs, but is also cleaner. If the slices/maps are returned without any users in them they will always be returned as empty slices/maps instead of nil values.